contract/src/offer-up.contract.js

```diff
@@ -1,70 +1,143 @@
 
-const trace = makeTracer('OfferUp', true);
+// #region bag utilities
+/** @type { (xs: bigint[]) => bigint } */
+const sum = xs => xs.reduce((acc, x) => acc + x, 0n);
 
-/** @param {Amount<'copyBag'>} amt */
-const bagValueSize = amt => {
-  /** @type {[unknown, bigint][]} */
-  const entries = getCopyBagEntries(amt.value); // XXX getCopyBagEntries returns any???
-  const total = entries.reduce((acc, [_, qty]) => acc + qty, 0n);
-  return total;
+/**
+ * @param {import('@endo/patterns').CopyBag} bag
+ * @returns {bigint[]}
+ */
+const bagCounts = bag => {
+  const entries = getCopyBagEntries(bag);
+  return entries.map(([_k, ct]) => ct);
 };
+// #endregion
 
 /**
- * @param {ZCF<{tradePrice: Amount}>} zcf
+ * This contract is parameterized by terms for price and,
+ * optionally, a maximum number of items sold for that price (default: 3).
+ *
+ * @typedef {{
+ *   tradePrice: Amount;
+ *   maxItems?: bigint;
+ * }} ItemTerms
+ */
+
+export const meta = {
+  customTermsShape: M.splitRecord(
+    { tradePrice: AmountShape },
+    { maxItems: M.bigint() },
+  ),
+};
+// compatibility with an earlier contract metadata API
+export const customTermsShape = meta.customTermsShape;
+
+/**
+ * Start a contract that
+ *   - creates a new non-fungible asset type for Items, and
+ *   - handles offers to buy up to `maxItems` items at a time.
+ *
+ * As is typical in Zoe contracts, the flow is:
+ *   1. contract does internal setup
+ *   2. client uses a public facet method -- {@link makeTradeInvitation} in this case --
+ *      to make an invitation.
+ *   3. client makes an offer using the invitation, along with
+ *      a proposal (with give and want) and payments. Zoe escrows the payments, and then
+ *   4. Zoe invokes the offer handler specified in step 2 -- here {@link tradeHandler}.
+ *
+ * @param {ZCF<ItemTerms>} zcf
  */
 export const start = async zcf => {
-  const { tradePrice } = zcf.getTerms();
+  const { tradePrice, maxItems = 3n } = zcf.getTerms();
 
-  const { zcfSeat: proceeds } = zcf.makeEmptySeatKit();
-  const mint = await zcf.makeZCFMint('Item', AssetKind.COPY_BAG);
+  /**
+   * a new ERTP mint for items, accessed thru the Zoe Contract Facet.
+   * Note: `makeZCFMint` makes the associated brand and issuer available
+   * in the contract's terms.
+   *
+   * AssetKind.COPY_BAG can express non-fungible (or rather: semi-fungible)
+   * amounts such as: 3 potions and 1 map.
+   */
+  const itemMint = await zcf.makeZCFMint('Item', AssetKind.COPY_BAG);
+  const { brand: itemBrand } = itemMint.getIssuerRecord();
 
-  const tradeShape = harden({
-    give: { Price: AmountShape },
-    want: { Items: AmountShape },
+  /**
+   * a pattern to constrain proposals given to {@link tradeHandler}
+   *
+   * The `Price` amount must use the brand from the `price` term.
+   * The `Items` amount must use the `Item` brand and a bag value.
+   */
+  const proposalShape = harden({
+    give: { Price: { brand: tradePrice.brand, value: M.any() } },
```

Comment:
I think you can use `M.gte()` here and not need `AmountMath.isGTE()` below
```suggestion
    give: { Price: { brand: tradePrice.brand, value: M.gte(price.amount) } },
```
or
```suggestion
    give: { Price: M.gte(price) },
```

Reply:
good point!